src/providers/baremetal.rs

```diff
@@ -0,0 +1,221 @@
+use crate::ssh;
+use failure::Error;
+use std::collections::HashMap;
+use std::sync::Arc;
+
+/// Descriptor for a single, existing machine to connect to.
+/// Therefore, the `impl MachineSetup` includes the address of the machine in `region`; i.e.,
+/// each instance of Setup corresponds to a single machine.
+#[derive(Clone)]
+pub struct Setup {
+    addr: std::net::SocketAddr,
+    username: String,
+    key_path: Option<std::path::PathBuf>,
+    setup_fn:
+        Option<Arc<dyn Fn(&mut ssh::Session, &slog::Logger) -> Result<(), Error> + Send + Sync>>,
+}
+
+impl super::MachineSetup for Setup {
+    type Region = String;
+    fn region(&self) -> Self::Region {
+        format!("bare:{}", self.addr)
+    }
+}
+
+impl PartialEq for Setup {
+    fn eq(&self, other: &Self) -> bool {
+        self.addr == other.addr
+    }
+}
+
+impl Eq for Setup {}
+
+impl std::hash::Hash for Setup {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.addr.hash(state);
+    }
+}
+
+impl Setup {
+    pub fn new(
+        addr: impl std::net::ToSocketAddrs,
+        username: Option<String>,
+    ) -> Result<Self, Error> {
+        let username: Result<String, Error> = username.map(Ok).unwrap_or_else(|| {
+            let user = String::from_utf8(std::process::Command::new("whoami").output()?.stdout)?;
+            let user = user
+                .split_whitespace()
+                .next()
+                .expect("expect newline after whoami output");
+            Ok(user.to_string())
+        });
+        let username = username?;
+
+        Ok(Self {
+            username,
+            addr: addr
+                .to_socket_addrs()?
+                .next()
```

Comment:
I think we may want to keep all the addrs around so that we can try each in turn if need be.